### PR TITLE
[WIP] feat: add cache provider implementation

### PR DESCRIPTION
**Pull request by milosgajdos**
Currently, the only two implementations of cache are `redis` and `inmemory`. 

This PR introduces cache providers that let you implement your own `blobdescriptor` cache of choice.

Extended rebase of https://github.com/distribution/distribution/pull/2580

**EDIT:** 

⚠️ **This changes configuration but it's still the perfect time to do this before v3 release** ⚠️ 

`cache` now has a dedicated configuration struct instead of being part of the storage configuration. 
This simplifies the code quite a bit at the expense of sad configuration decoding. I passionately hat the `configuration` package and IMHO we should rework it almost from the ground up.

Each cache provider must register itself via the`init()` func so it's loaded on startup and available if the config requires it.

Adding new cache providers, should there be such a need in the future, should be easier. 
Currently, we continue to provide only `redis` and `inmemory` cache provider implementations.

Cache configuration is now expected to be set with the following options:
* `provider`: name of a cache provider
* `params` set of config params that configure the given cache provider

`Redis` configuration has been moved from the main configuration to `redis` cache pacakge because its only ever been used for configuring caches and nothing else so it makes no sense to keep it in the global config.

Documentation has been updated accordingly.

### CODE REVIEW

File: registry/storage/cache/provider/cacheprovider.go
Comment (corhere) [inline registry/storage/cache/provider/cacheprovider.go:24]: Is `provider.Register` intended to be called from `init()` functions for packages to self-register? If so, why not just panic?

File: registry/storage/cache/provider/cacheprovider.go
Comment (corhere) [inline registry/storage/cache/provider/cacheprovider.go:12]: Pet peeve: just call it what it is.
```suggestion
type Factory func(ctx context.Context, options map[string]interface{}) (cache.BlobDescriptorCacheProvider, error)
```

File: registry/storage/cache/provider/cacheprovider.go
Comment (corhere) [inline registry/storage/cache/provider/cacheprovider.go:37]:
```suggestion
	if initFunc, exists := cacheProviders[name]; exists {
		return initFunc(ctx, options)
	}
```

File: registry/storage/cache/provider/cacheprovider.go
Comment (corhere) [inline registry/storage/cache/provider/cacheprovider.go:1]: I wonder how many near-duplicates of this pattern are scattered throughout the codebase. I betcha they could all be replaced with instantiations of a single generic implementation.
```go
type FactoryRegistry[T any, F func(context.Context, map[string]any] struct{ ... }
func NewFactoryRegistry[T any, F func(context.Context, map[string]any) (T, error)]() FactoryRegistry[T]

func (*FactoryRegistry[T, F]) Register(name string, factory F)
func (*FactoryRegistry[T, F]) Get(ctx context.Context, name string, options map[string]any) (T, error)
```

```go
var Registry = factoryutil.NewFactoryRegistry[cache.BlobDescriptorCacheProvider]()
```

File: registry/storage/cache/provider/cacheprovider.go
Comment (milosgajdos) [reply registry/storage/cache/provider/cacheprovider.go:12]: this is called `InitFunc` because that's what we call similar funcs throughout the codebase e.g
* https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/middleware/storagemiddleware.go#L12
* https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/middleware/registry/middleware.go#L14

I'm not a major fan of it either but I think we should stay consistent

Comment (corhere) [reply registry/storage/cache/provider/cacheprovider.go:12]: Fair enough.

File: registry/storage/cache/provider/cacheprovider.go
Comment (milosgajdos) [reply registry/storage/cache/provider/cacheprovider.go:1]: I think with generics, this codebase could use proper massaging.

File: registry/storage/cache/provider/cacheprovider.go
Comment (milosgajdos) [reply registry/storage/cache/provider/cacheprovider.go:24]: Good shout. I think there are a few other cases in this codebase where we should rectify this e.g. https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/middleware/storagemiddleware.go#L18-L29

**Review by corhere — DISMISSED** [review]

**Comment by milosgajdos** [comment]
@corhere I've worked the cache provider stuff from the ground up. Would love it if you could spare a minute looking through this PR. I've marked it as a `draft` for the time being so we can iron out some details, but I'm pretty happy with it as it is.

Also PTAL @thaJeztah @Jamstah

**Review by corhere — CHANGES_REQUESTED** [review]
You might be able to convince me regarding the change to redis configuration.

File: registry/storage/cache/memory/memory.go
Comment (milosgajdos) [reply registry/storage/cache/memory/memory.go:185]: Yes, it's one of a few things in this PR that I'm still mulling over. Does not feel right I agree. 

Note that the Azure storage driver is also decoding auth config in its constructor: thats where I got some inspiration, but I'm not happy about it no.

It's also one of the reasons I marked this PR as draft. There are things to be figured out here.

File: cmd/registry/config-cache.yml
Comment (milosgajdos) [reply cmd/registry/config-cache.yml:30]: I was originally thinking about EXACTLY the same lines before I made the change.

The use cases you mention might make some sense in the future but I'm skeptical: nobodoy will implement rate limiters into this bloating codebase. Large deployments are usually behind proxies that handle rate limiting before the requests hit the API -- speaking of experience of managing many of these large volume ones over my career.

Whilst sharing a Redis instance across many use cases makes sense in cute little deployments it becomes a nuisance and proper headache whem you outgrow your small setup, not to mention whatever security use case your security team barfs at you, etc.

Still Not entirely sold on keeping it as a global, but I'm also not strictly against it either - just mulling over my past experiences. 

Sometimes it's better to duplicate things than create more problems for yourself by prematurely optimising 🙃

File: registry/storage/cache/memory/memory.go
Comment (milosgajdos) [reply registry/storage/cache/memory/memory.go:185]: Just FYI: this is a helper func, you don't have to use it, there is nothing that prevents you from providing your own `map[string]interface{}` params in a similar way it's done in the storage drivers.

I was trying to avoid this (and other storage driver config) absolute state of things https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/s3-aws/s3.go#L192-L472

But maybe I'm trying to be too clever here 🤷‍♂️

File: registry/storage/cache/cache.go
Comment (milosgajdos) [reply registry/storage/cache/cache.go:47]: Yeah, this is possible, but I'm trying to follow the established conventions in this codebase (see the storage drives and middleware)

File: registry/storage/cache/memory/memory.go
Comment (corhere) [reply registry/storage/cache/memory/memory.go:185]: Code needs to exist somewhere which transforms a generic `map[string]any` configuration blob into something the implementation can consume. There's no getting away from the code in the body of `s3.FromParameters` aside from leveraging metaprogramming (e.g. `mapstructure`) to do the same thing less verbosely. The only notes I have for `s3.FromParameters` are:
1. It would be easier to unit-test if the code to parse `parameters` was extracted to its own `func(map[string]any) (DriverParamters, error)`
2. `FromParameters` probably does not need to be exported since it is made available through the driver registry. Code which does not need to go through the registry **and therefore does not need to be generic over all storage drivers** does not need to go through the generic parameter-parsing machinery.
3. The functional options pattern would make it easier to evolve the configuration and implementation in semver-compatible ways than to export the `DriverParameters` struct.

Aside from that, I firmly believe that the cache provider implementations _should_ be following the way the s3 storage driver does configuration and construction. Consumers that do not need to be generic over all implementations of an interface should not be forced to use the weakly-typed, generic-over-all-implementations factory interface targeted at humans. Configuration as `map[string]any` makes runtime errors out of what could have otherwise been caught at compile time.

File: registry/storage/cache/cache.go
Comment (corhere) [reply registry/storage/cache/cache.go:47]: What's so sacred about the established conventions?

Comment (milosgajdos) [reply registry/storage/cache/cache.go:47]: Nothing in particular, but I prefer keeping code consistent across codebases: it makes navigation, comprehension, onboarding new devs, etc much easier...I got a bit religious with it in this codebase I agree

Comment (deleteriousEffect) [reply registry/storage/cache/cache.go:47]: I agree for keeping code consistent, no code is perfect and having to deal with the same suboptimal convention everywhere is better than having to deal with 5 or 6 practices of varying quality, even if the average quality is better. 

If we did want to establish a new convention, we should be deliberate about picking one and moving forward with that single practice, and we should also look towards switching the old code over to that as well.

We don't have to rip everything out all at once, but if the new convention is not compelling enough that we'd see benefit from applying it to the older code, then it may also not be compelling enough to deal with two conventions during the interim period.

Comment (corhere) [reply registry/storage/cache/cache.go:47]: Let me get this straight:
- All existing code has been grandfathered in as "conventional."
  - Except for the S3 storage driver. For some reason, everything it does has been decreed anti-conventional.
- All new code to which any similarity can be drawn with existing code must be shaped like the existing code.
  - Except for the S3 storage driver. [New code must be _inconsistent with_ the S3 driver, just to be contrary.](https://github.com/distribution/distribution/pull/4169#discussion_r1421043575)
- Any changes to how code is shaped must be agreed upon by consensus and must then be ratified by applying it to the entire codebase.
  - Any changes must therefore go through a waterfall process before a single line of new code can be merged which goes against grandfathered conventions.
  - The determination of whether a new convention is compelling enough must be made without having experience coding in the new convention. Experience in coding the new convention cannot be gained without coding in the new convention. New conventions are therefore de facto banned. Whatever conventions were grandfathered in are our coding conventions, now and forever.
- And given which thread this discussion is being made on, anything which might possibly _afford_ code to be written in a divergent style _while still affording the conventional style_ is **also banned**

That can't be right.

I am literally just suggesting that the _brand new_ `Register` function grow a dummy result parameter so that implementations _could_ use a less-noisy syntax for registration. The following would also compile just fine!
```go
func init() {
        cache.Register("myprovider", initFunc)
}
```

I'm here to push boundaries and shake up existing conventions in order to evolve the codebase. If existing conventions are so sacred that they cannot even be iterated upon in greenfield code, I do not see a future where I remain on as a distribution maintainer.

Comment (milosgajdos) [reply registry/storage/cache/cache.go:47]: > I'm here to push boundaries and shake up existing conventions in order to evolve the codebase. If existing conventions are so sacred that they cannot even be iterated upon in greenfield code, I do not see a future where I remain on as a distribution maintainer.

FYI: I appreciate your feedback very much. In fact, I really enjoy working together/collaborating on this project with you. I had no idea some of my comments would cause this much discussion or maybe even angst (hopefully not!).

This codebase does need a ton of improvements all around and I need to withhold [at least some of my] hot takes (e.g. "set it on fire one lol") here as they're not productive. This discussion should probably belong elsewhere though.

FYI2: I'm not against changing the `InitFunc` - I'm sure we can find a way to make this work 😄

File: registry/storage/cache/memory/memory.go
Comment (milosgajdos) [reply registry/storage/cache/memory/memory.go:40]: Some open questions I still have (besides the cohesion point you highlighted -- thanks!):
* should we move `redis` config stanza out of the root of the config -- I'm undecided on this; I feel like having it in the root makes it look like Redis is an "essential" part of the app which isn't true: we can instead either use in memory cache or no cache at all. So creating some semantic section for it that highlights its purpose in the wider context of the app could make sense. No idea. I'm just braindumping here.
* What should the params passed down to Redis module look like? We've agreed on config decoder living outside of it. To make it easier to consume Redis cache APIs should the `pool` be created inside the constructor from the passed-in params or should it be created in the app root (this would allow keeping the global pool client in the app and potentially reusing it elsewhere) and then passed it via `params` to the module (e.g. `map["pool"] = pool` - this was actually my original idea, but it didn't feel quite right). The former would align with _some_ of the code in this project but make it mildly annoying to consume (i.e. the API consumers would have to supply parameters via the mildly type unsafe `map` -- `any` values gonna be `any`ing), the latter would make it easier to consumer (i.e. the API consumers could simply just pass the pool object down) but it doesn't quite feel right either; the last option could be changing the `InitFunc` that could maybe accept the `pool` as param directly.

I'm still mulling over this. And since this PR is still a WIP or more like an exercise in exploration of the future, there is no rush on this. I'd prefer getting this right (or as close to the ideal as possible) than regret it in the future

File: registry/storage/cache/cache.go
Comment (deleteriousEffect) [reply registry/storage/cache/cache.go:47]: > I'm here to push boundaries and shake up existing conventions in order to evolve the codebase. If existing conventions are so sacred that they cannot even be iterated upon in greenfield code, I do not see a future where I remain on as a distribution maintainer.

Please know that I am also intensely frustrated by the way this codebase does things. This project is old enough to have existed before a lot of Go best practices were solidified, and it shows.

> Any changes to how code is shaped must be agreed upon by consensus

You need two approvals from two maintainers in different orgs, in your case that's yourself and one other maintainer.

I tend to be more cautious, and I think it's a good and healthy balance to have someone who wants to shake things up. We need to moderate each other in this way. 

> and must then be ratified by applying it to the entire codebase.

I have a solution below that might give us both consensus and the capacity to effect change.

>     The determination of whether a new convention is compelling enough must be made without having experience coding in the new convention. Experience in coding the new convention cannot be gained without coding in the new convention. New conventions are therefore de facto banned. Whatever conventions were grandfathered in are our coding conventions, now and forever.

We all have worked on other projects and seen different conventions applied there and what their results were.
We can reasonably determine if something is going to work out and then decide to commit to it. 

If we want to adopt a new convention, we can start with a single PR. However, if there's no barrier to entry, then it's easy to have a bunch of unrelated incohesive practices sprinkled throughout the code. Part of being a maintainer on any project is thinking about the impact of changes in context of existing code. 

What I think would help in resolving these conflicts is adopting or implementing a style guide. This way, we can be deliberate about what our practices are, and apply them consistently. This would allow us to stabilize our practices moving forward and justify and encourage creating PRs which make old code conform. Having a style guide should provide the consistency that I want to see, while providing you with the ability to evolve the codebase.

To move forward, I propose:

- Bucking the convention of returning interfaces, right now, in this PR. 
- Creating a style guide with a single entry in support of this style, in this or a followup PR.
- Creating an issue to discuss expanding the style guide, for example, if we want to write our own or adopt some existing guide and if so, which one.

File: registry/storage/cache/memory/memory.go
Comment (corhere) [reply registry/storage/cache/memory/memory.go:40]: > * should we move `redis` config stanza out of the root of the config

I am not too concerned about confusing users by keeping the config stanza in the root. I still think there is some merit to future-proofing the config to afford sharing a single redis client connection pool (which implies a shared database instance) among multiple redis-backed modules within the app. But I don't feel too strongly about it: if in the short-term it would require a complicated architecture to pull it off when there is just a single consumer, I am fully prepared to accept that YAGNI.

(Aside: in a hypothetical future where there is more than one redis-backed module available in the app, root redis config could coexist with per-module overrides. The overridden module would instantiate its own client, inheriting the root redis config values as defaults. E.g.:)
```yaml
redis:
  addr: localhost:6379
  pool:
    maxidle: 16
    maxactive: 64
    idletimeout: 300s
  dialtimeout: 10ms
  readtimeout: 10ms
  writetimeout: 10ms
cache:
  blobdescriptor:
    provider: redis
      params:
        db: 42
```

> * What should the params passed down to Redis module look like?

The _constructor_ should continue to accept a `*redis.Client` directly. How the _factory_ function acquires that client is the big question.

First thing we could do is move the cache-provider registry stuff under an `internal` subtree to free it from the SemVer compatibility contract. (Maybe the cache implementations, too? With sanely-shaped constructors they _might_ be useful for third parties, but we'd have to think about SemVer compatibility.) Then we'd be free to iterate on the registry and factories without being constrained by past decisions.

In the long term, a shared app-wide redis client (for example) could be made available to the cache-provider factories through some form of dependency injection. A pointer to the app struct could be passed into `registry.Get` alongside the params map, or perhaps some purpose-built "provider registry" object. In the short term, said dependency injection mechanism could instead provide the top-level redis config to the factory since the redis cache-provider is the only thing which needs the redis client today. With `app.redis` already being a thing, passing the app struct pointer into the factory may be the path of least resistance as long as we are okay with the factory implementation living in the `app` package. (I think it makes sense: the factory is necessarily tightly coupled to (at least one subtree of) the app config so it follows logically that it should also have high cohesion with the app package.)

**Comment by milosgajdos** [comment]
Closing as not planned. It was an experiment aiming at cleaning up the cache config cruft.

**Pull request closed without merging**
